### ISSUE TITLE
Poetry shell seems to spawn a non-login shell, giving a very minimalistic prompt

### ISSUE THREAD

**Issue by stefanoborini** [python-poetry/poetry#1919]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.

- **OS version and name**: macOS 10.14.6
- **Poetry version**: 1.0.2

## Issue

``poetry shell`` seems to override the prompt completely in a very aggressive way. My prompt currently looks like this

```
username@host:mydir (gitrepo) [15 stash] $
```

When I invoke ``poetry shell`` the resulting prompt looks like

```
Spawning shell within .../virtualenvs/omitted
bash-3.2$ . /omitted/bin/activate
(rip-oEcYTxbK-py3.7) bash-3.2$ cd -
```

I suspect this is because the shell is not spawned as a login shell, hence it does not parse configuration files (e.g. bashrc). This might be a bug or a missing feature. If the second, ``poetry shell`` should have an option to spawn a login shell.

**Comment by rcludwick** [comment]
I think this is worse than this.  

Running poetry shell for bash clears anything defined in .bash_profile, which is where the standard "run once" interactive shell definitions go.   Anything setup like development variables, custom prompts all get wiped away.

It will read .bashrc though.  

This seems to suggest that current environment variables set in the current shell, might also get blown away.

Note this behavior is different than running activate inside the virtualenv because activate doesn't blow away your shell environment.

Note that pipenv shell doesn't blow away the definitions set by .bash_profile.

**Comment by d3QUone** [comment]
Same for me:
* OS version and name: macOS 10.15.5
* Poetry version 1.0.9

**Comment by ericsouza** [comment]
Same problem, any suggestion?

**Comment by aprepo** [comment]
Same problem, any workaround?

**Comment by erny** [comment]
It seems that poetry (mine: 1.1.6) on MacOS (mine: 10.14.6) ignores .bash_profile but sources .bashrc. So the easy solution is to create a ~/.bashrc with the line:

    . ~/.bash_profile

This allows you to comment why the file (.bashrc) is there and additional commands.

**Comment by rcludwick** [comment]
My work around is to not bother with poetry shell.

**Comment by rcludwick** [comment]
> It seems that poetry (mine: 1.1.6) on MacOS (mine: 10.14.6) ignores .bash_profile but sources .bashrc. So the easy solution is to create a ~/.bashrc with the line:
> 
> ```
> . ~/.bash_profile
> ```
> 
> This allows you to comment why the file (.bashrc) is there and additional commands.

So I'm a gray beard now (I wasn't 20 years ago weirdly) and this a completely unacceptable solution.  

PATH settings need to go in .bash_profile, so they're read exactly once by the login shell.  If you source .bash_profile from 
  .bashrc, then every subsequent bash shell (including those in scripts) will get it's PATH modified before it runs.  

Ex: 

.bashrc: 
export PATH=/usr/local/bin:$PATH

Then you do:  
$ export PATH=/bin:/usr/bin
$ bash
$ echo $PATH
$ bash
$ echo $PATH

It's a f'ing nightmare to debug if you're trying to figure out why some script is running an executable from a directory it shouldn't be.

That's why .bash_profile exists.

**Comment by erny** [comment]
Thank you for the feedback. What would be an aceptable solution? When I'm running poetry I'm getting a complete clean environment. No aliases, no PS1, no PATHs.

Regards.

**Comment by erny** [comment]
Related ticket #4180

**Comment by rcludwick** [comment]
Ideally poetry shouldn't fork off another shell when activation of an existing virtualenv is all that's needed.  

1.  Use poetry config to set the venv elsewhere that's really handy.   I have my venvs live in ~/venv. 
2.  Source the activate script from the virtualenv (. ~/venv/project/bin/activate) instead of poetry shell. 
3.  Live in peace knowing that your environment has been treated with all the proper respect it should be.

**Comment by djmattyg007** [comment]
Why not just run `$SHELL -l` and spawn a login shell? Or at least give us the option as a flag to `poetry shell`? It could just be `poetry shell -l`.

**Comment by LeonardoGentile** [comment]
Any update on this?

**Comment by neersighted** [comment]
This is mostly a user configuration issue -- interactive non-login shells are a routine part of user life and if you want all your customizations to be present you should read your shell's documentation./ That being said, a `poetry shell -l` feature might be desirable -- MRs are welcome, as well as a new issue tracking that feature request.

I'm going to close this for now as in general there's not much Poetry can do about your shell config -- please open an issue for the suggested `shell -l` if you are interested.

**Comment by rcludwick** [comment]
When you obliterate the environment as you do currently during poetry shell,  you are deciding for the user what the environment should be.  

That is the issue.

Don't blame the issue on bash misconfiguration.  Here's what the manpage says on bash:

>        When bash is invoked as an interactive login shell, or as a  non-inter‐
>        active  shell with the --login option, it first reads and executes com‐
>        mands from the file /etc/profile, if that file exists.   After  reading
>        that file, it looks for ~/.bash_profile, ~/.bash_login, and ~/.profile,
>        in that order, and reads and executes commands from the first one  that
>        exists  and  is  readable.  The --noprofile option may be used when the
>        shell is started to inhibit this behavior.

Setting PATH once and only once is a reasonable thing to do in the .bash_profile.  Subsequent bash interactive shells should never ever have to touch PATH ever again.  

When poetry shell runs, this gets blown away because you chose for it to get blown away.   Whatever PATH I set is now gone.  Hence the issue.

**Comment by neersighted** [comment]
There is not another way for us to 'activate' a virtual environment for the user -- our options are 
1. write a script users can source to stdout
2. spawn a subshell and feed a command to activate to stdin

The correct solution to shell customization is some combination of login and interactive settings, or using both `.bashrc` and `.bash_profile`. Poetry is not doing anything to change your path -- `poetry shell` is the same as `bash` and then typing `source /path/to/venv/bin/activate`. If your path is clobbered by an interactive non-login shell, that is entirely a configuration issue.

It is possible for us to not 'type' `source` for the user and to directly inject `PATH` and `VIRTUAL_ENV` into the child process we spawn (see `pipenv shell --fancy`), but this actually is closer to what you are complaining about and will lead to frequent support issues for users with misconfigured shells (see: the pipenv project's experience).

In short, please explain how what Poetry does is any different from `bash` and `source activate` -- I know that login shells are the default on macOS for new terminals, but on any other Unix platform only the root shell of a login session is login by default, and you must customize your shell with this in mind.

**Comment by rcludwick** [comment]
Okay I did a review of the poetry code in this area, and it looks fine to
me.

I saw this commit, which may explain part of what I was seeing.  It looked
like back then, poetry was mutating PATH directly.

https://github.com/python-poetry/poetry/commit/15d7a263b2cde791823db168cb17d700b490f534

I'm happy now.  This can stay closed.

**Comment by willmurphyscode** [comment]
Thanks for the detailed discussions on this issue, and thanks for poetry! I'm pretty new to python, and it's nicer than other things I've tried. 

I found this issue when searching for why poetry shells had some different behavior from regular shells, and it led me here. Wanted to post what's worked for me (zsh on m1 mac, if that matters), in case others find this issue too. 

TL;DR: I moved the following stuff from `~/.zprofile` to `~/.zshrc`, and now poetry shells work just like I'd hoped:

``` zsh
# (all my aliases)

# activate z utility to jump around between directories
. $HOME/.z_util/z.sh

# set up fzf for nice ctrl+r history searching
[ -f ~/.fzf.zsh ] && source ~/.fzf.zsh

# re-load completions in poetry shells
autoload -Uz compinit && compinit
```

@rcludwick and @erny pointed out that sourcing a profile file from an RC file can result in problems like repeatedly appending to PATH, etc. But PATH isn't what we need to change.

``` sh
$ echo $PATH | tr ':' '\n' > /tmp/regular-path
$ echo $PATH | tr ':' '\n' > /tmp/poetry-shell-path
$ diff /tmp/poetry-shell-path /tmp/regular-path
1d0
< /Users/willmurphy/Library/Caches/pypoetry/virtualenvs/vunnel-8rbbXRxd-py3.11/bin
```

You can see that `poetry shell` isn't messing up PATH at all; it's just putting the virtual env's bin folder first on PATH.

So I think we can all be happy by putting things like aliases and autocomplete setups in our `~/zshrc` or `~/.bashrc` file, and leaving one-time PATH manipulation in a profile file. Hope this helps someone!
